steely/plugins/_haskell.sh

```diff
@@ -0,0 +1,43 @@
+DIRECTORY="temporary"
+EXECUTABLE="$DIRECTORY/main"
+FILENAME="$DIRECTORY/code.hs"
+
+function _clean_up {
+    rm -rf $DIRECTORY
+    mkdir $DIRECTORY
+}
+
+function _save {
+    code=$(cat)
+    echo "$code" > $FILENAME
+}
+
+function _compile {
+    echo
+    echo "*Compiler Output:*"
+    ghc -o $EXECUTABLE $FILENAME
+}
+
+function _run {
+    echo
+    echo "*Program Output:*"
+    local secondsToWait=4
+    local whitelist="$(pwd)/$DIRECTORY"
+
+    timeout $secondsToWait firejail --quiet --noprofile --whitelist=$whitelist ./$EXECUTABLE || _checkTimeout
+}
+
+function _checkTimeout {
+    local exitCode=$?
+    if [[ $exitCode -ne 0 ]]; then
+        echo "Program timed out. Don't be cheeky."
```

Comment:
I don't think it's possible for _checkTimeout to be called, since it's only called if time out exits with a non zero

Reply:
> `_checkTimeout` is only called if time out exits with a non zero.

Yes, that's intended.

Because I'm using `set -e` (exit when there's an error), I need to use `||` to check if there was a timeout (and display a nice message).

It works, as shown in the yolanda chat.

Reply:
Is it necessary to check the exit code if you already know it's non-zero?

Reply:
Yes. A non-zero exit code from timeout doesn't mean that it timed out.

From `man timeout`:
```
If the command times out, and --preserve-status
is not set, then exit with status 124. Otherwise,
exit with the status of COMMAND.
```

Reply:
Oops you're right. I need to check for 124. I forgot I had the `-ne 0`.